### PR TITLE
refactor: rewrite Search with hook

### PR DESCRIPTION
**Pull request by 0xPoe**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Perfermance optimization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link

None.

### 💡 Background and solution

Rewrite Search with hook.

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |      refactor:  rewrite Search with hook     |
| 🇨🇳 Chinese |     refactor:  用 hook 重写 search 组件        |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by Rustin-Liu** [comment]
@afc163 @zombieJ 我尝试加了下这部分逻辑的测试，但是没能找到一个比较好的方法测试到。
我直接去使用 document.activeElement 似乎不生效，永远 focus 在 body 上。
你们有什么建议吗？或者有可以参考的例子和思路吗？ 谢谢！

**Comment by afc163** [comment]
试一下测试的例子挂载到 document 上。https://github.com/ant-design/ant-design/blob/02d729257699a053178d1322541afce894af98f5/components/input/__tests__/index.test.js#L180

**Comment by Rustin-Liu** [comment]
> 试一下测试的例子挂载到 document 上。

昨晚试了，还是一直在 body 上。但是直接挂个 input 上去，然后直接去 fcous() 就可以获取焦点，我们渲染的组件就不行，但实际是 focus 的 mock 已经被触发了。就很奇怪。

**Comment by Rustin-Liu** [comment]
@zombieJ  @afc163  @xrkffgg  Pass all tests. PTAL~

File: components/input/Search.tsx [hunk 1]
```diff
@@ -199,22 +183,26 @@ export default class Search extends React.Component<SearchProps, any> {
...
-  render() {
-    return <ConfigConsumer>{this.renderSearch}</ConfigConsumer>;
-  }
-}
+  return <ConfigConsumer>{renderSearch}</ConfigConsumer>;
+});
+
+Search.defaultProps = {
+  enterButton: false,
+};
```
Comment (afc163) [inline components/input/Search.tsx:206]: displayName 要补充一下。

**Pull request merged**